lib/run/task.ts

```diff
@@ -86,10 +115,37 @@ export function createTask<T>(
             });
           }
         },
-        then: (...args) => getHaltPromise().then(...args),
-        catch: (...args) => getHaltPromise().catch(...args),
-        finally: (...args) => getHaltPromise().finally(...args),
+        // then: (...args) => getHaltPromise().then(...args),
+        then: async <Result1=void, Result2=never>(onfulfilled?: ThenFulfilledType<void, Result1>, onrejected?: CatchType<Result2>) => {
+          type NewResult = Result1 | Result2;
+          const newPromise = getHaltPromise().then(onfulfilled, onrejected);
+          const future: Future<NewResult> = create<Future<NewResult>>("Future", {}, {
+            *[Symbol.iterator]() {
+              return yield* call(() => newPromise);
+            },
+            then: (...args) => newPromise.then(...args),
+            catch: (...args) => newPromise.catch(...args),
+            finally: (...args) => newPromise.finally(...args),
+          });
+          return await future;
```

Comment:
note: the code here for handling the abort future is mostly the same as the `then` in the parent `Task`

This is why I think maybe it's better if we have a more general way to create a `Future` that is exposed somewhere that is used in both these places